src/key_pair.rs

```diff
@@ -106,65 +105,110 @@ impl KeyPair {
-		let pkcs8_vec = pkcs8.to_vec();
+		if alg == &PKCS_ED25519 {
+			return Ok(Self::pkcs_ed25519(pkcs8)?);
+		}
+		if alg == &PKCS_ECDSA_P256_SHA256 {
+			return Ok(Self::pkcs_ecdsa_p256_sha256(pkcs8)?);
+		}
+		if alg == &PKCS_ECDSA_P384_SHA384 {
+			return Ok(Self::pkcs_ecdsa_p384_sha384(pkcs8)?);
+		}
+		if alg == &PKCS_RSA_SHA256 {
+			return Ok(Self::pkcs_rsa_sha256(pkcs8)?);
+		}
+		if alg == &PKCS_RSA_SHA384 {
+			return Ok(Self::pkcs_rsa_sha384(pkcs8)?);
+		}
+		if alg == &PKCS_RSA_SHA512 {
+			return Ok(Self::pkcs_rsa_sha512(pkcs8)?);
+		}
+		if alg == &PKCS_RSA_PSS_SHA256 {
+			return Ok(Self::pkcs_rsa_pss_sha256(pkcs8)?);
+		}
 
-		let kind = if alg == &PKCS_ED25519 {
-			KeyPairKind::Ed(Ed25519KeyPair::from_pkcs8_maybe_unchecked(pkcs8)?)
-		} else if alg == &PKCS_ECDSA_P256_SHA256 {
-			KeyPairKind::Ec(EcdsaKeyPair::from_pkcs8(
+		panic!("Unknown SignatureAlgorithm specified!")
+	}
+
+	pub(crate) fn guess_kind_from_der(der: &[u8]) -> Result<KeyPair, Error> {
+		if let Ok(kp) = Self::pkcs_ed25519(der) {
+			return Ok(kp);
+		}
+
+		if let Ok(kp) = Self::pkcs_ecdsa_p256_sha256(der) {
+			return Ok(kp);
+		}
+
+		if let Ok(kp) = Self::pkcs_ecdsa_p384_sha384(der) {
+			return Ok(kp);
+		}
+
+		if let Ok(kp) = Self::pkcs_rsa_sha256(der) {
+			return Ok(kp);
+		}
+
+		return Err(Error::CouldNotParseKeyPair);
+	}
+
+	fn pkcs_ed25519(der: &[u8]) -> Result<KeyPair, Error> {
+		Ok(KeyPair {
+			kind: KeyPairKind::Ed(Ed25519KeyPair::from_pkcs8_maybe_unchecked(der)?),
+			alg: &PKCS_ED25519,
+			serialized_der: der.to_vec(),
+		})
+	}
+
+	fn pkcs_ecdsa_p256_sha256(der: &[u8]) -> Result<KeyPair, Error> {
+		Ok(KeyPair {
+			kind: KeyPairKind::Ec(EcdsaKeyPair::from_pkcs8(
 				&signature::ECDSA_P256_SHA256_ASN1_SIGNING,
-				pkcs8,
-			)?)
-		} else if alg == &PKCS_ECDSA_P384_SHA384 {
-			KeyPairKind::Ec(EcdsaKeyPair::from_pkcs8(
+				der,
+			)?),
+			alg: &PKCS_ECDSA_P256_SHA256,
+			serialized_der: der.to_vec(),
+		})
+	}
+
+	fn pkcs_ecdsa_p384_sha384(der: &[u8]) -> Result<KeyPair, Error> {
+		Ok(KeyPair {
+			kind: KeyPairKind::Ec(EcdsaKeyPair::from_pkcs8(
 				&signature::ECDSA_P384_SHA384_ASN1_SIGNING,
-				pkcs8,
-			)?)
-		} else if alg == &PKCS_RSA_SHA256 {
-			let rsakp = RsaKeyPair::from_pkcs8(pkcs8)?;
-			KeyPairKind::Rsa(rsakp, &signature::RSA_PKCS1_SHA256)
-		} else if alg == &PKCS_RSA_SHA384 {
-			let rsakp = RsaKeyPair::from_pkcs8(pkcs8)?;
-			KeyPairKind::Rsa(rsakp, &signature::RSA_PKCS1_SHA384)
-		} else if alg == &PKCS_RSA_SHA512 {
-			let rsakp = RsaKeyPair::from_pkcs8(pkcs8)?;
-			KeyPairKind::Rsa(rsakp, &signature::RSA_PKCS1_SHA512)
-		} else if alg == &PKCS_RSA_PSS_SHA256 {
-			let rsakp = RsaKeyPair::from_pkcs8(pkcs8)?;
-			KeyPairKind::Rsa(rsakp, &signature::RSA_PSS_SHA256)
-		} else {
-			panic!("Unknown SignatureAlgorithm specified!");
-		};
+				der,
+			)?),
+			alg: &PKCS_ECDSA_P384_SHA384,
+			serialized_der: der.to_vec(),
+		})
+	}
 
+	fn pkcs_rsa_sha256(der: &[u8]) -> Result<KeyPair, Error> {
```

Comment:
Could you remove these functions in favour of the old state which assigned `kind`? This saved us a lot of repetitive code.